gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/PartitionedRegion.java

```diff
@@ -11888,34 +11888,57 @@ public void clearLocalPrimaries() {
       // (probably not required to do this in loop after the recovery lock)
       // [sumedh] do we need both recovery lock and bucket locks?
       boolean done = false;
-      Set<BucketRegion> lockedRegions = null;
+      final ArrayList<BucketRegion> lockedRegions = new ArrayList<>();
       while (!done) {
-        lockedRegions = getDataStore().getAllLocalPrimaryBucketRegions();
+        // release locks on any buckets locked in previous iteration
+        if (!lockedRegions.isEmpty()) {
```

Comment:
Why is it necessary to unlock earlier locked regions? We can simply skip the regions while taking locks in subsequent attempts.

Reply:
Ideally the locking here should be in a consistent order to avoid any possible deadlocks between two truncates, for example, so releasing is best. Though the getAllLocalPrimaryBucketRegions() returns a Set is not proper, so will change it to return an ordered collection instead.